[Bug] [WorkflowRunningStateAction] Killed state of timeout failure strategy will cause workflow can't finish or stop

### Search before asking

- [x] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar issues.


### What happened

when a task was auto killed by timeout strategy, `WorkflowRunningStateAction.emitWorkflowFinishedEventIfApplicable` should check if there is no failed task and only exist killed or succeed tasks to finish the workflow.

```
[WI-4882819][TI-0] - 2025-07-21 00:36:01.313 INFO  [ds-workflow-eventbus-worker-15] o.a.d.s.m.e.t.l.h.TaskTimeoutLifecycleEventHandler:[71] - The task Task-Of-A TimeoutStrategy is FAILED, try to publish a kill event.
[WI-4882819][TI-0] - 2025-07-21 00:36:01.313 INFO  [ds-workflow-eventbus-worker-15] o.a.d.s.m.e.WorkflowEventBus:[41] - Publish event: TaskKillLifecycleEvent{task=Task-Of-A, delayTime=0}
```

```
[WI-4882819][TI-0] - 2025-07-21 06:15:18.149 ERROR [ds-workflow-eventbus-worker-12] o.a.d.s.m.e.WorkflowEventBusFireWorker:[88] - Fire event failed for WorkflowExecuteRunnable: WORKFLOW-A-20250721003500325
org.apache.dolphinscheduler.server.master.engine.exceptions.WorkflowEventFireException: Failed to fire event: WorkflowTopologyLogicalTransitionWithTaskFinishLifecycleEvent{task=[xlsx]1104_reporttaskState=SUCCESS}
        at org.apache.dolphinscheduler.server.master.engine.WorkflowEventBusFireWorker.doFireSingleWorkflowEventBus(WorkflowEventBusFireWorker.java:133)
        at org.apache.dolphinscheduler.server.master.engine.WorkflowEventBusFireWorker.fireAllRegisteredEvent(WorkflowEventBusFireWorker.java:86)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalStateException: The workflow: WORKFLOW-A-20250721003500325 state is RUNNING_EXECUTION can only finish with success/failed but exist task which state is not success and failure
        at org.apache.dolphinscheduler.server.master.engine.workflow.statemachine.WorkflowRunningStateAction.emitWorkflowFinishedEventIfApplicable(WorkflowRunningStateAction.java:151)
        at org.apache.dolphinscheduler.server.master.engine.workflow.statemachine.AbstractWorkflowStateAction.tryToTriggerSuccessorsAfterTaskFinish(AbstractWorkflowStateAction.java:126)
        at org.apache.dolphinscheduler.server.master.engine.workflow.statemachine.WorkflowRunningStateAction.topologyLogicalTransitionEventAction(WorkflowRunningStateAction.java:56)
        at org.apache.dolphinscheduler.server.master.engine.workflow.lifecycle.handler.WorkflowTopologyLogicalTransitionWithTaskFinishLifecycleEventHandler.handle(WorkflowTopologyLogicalTransitionWithTaskFinishLifecycleEventHandler.java:38)
        at org.apache.dolphinscheduler.server.master.engine.workflow.lifecycle.handler.WorkflowTopologyLogicalTransitionWithTaskFinishLifecycleEventHandler.handle(WorkflowTopologyLogicalTransitionWithTaskFinishLifecycleEventHandler.java:28)
        at org.apache.dolphinscheduler.server.master.engine.workflow.lifecycle.handler.AbstractWorkflowLifecycleEventHandler.handle(AbstractWorkflowLifecycleEventHandler.java:51)
        at org.apache.dolphinscheduler.server.master.engine.workflow.lifecycle.handler.AbstractWorkflowLifecycleEventHandler.handle(AbstractWorkflowLifecycleEventHandler.java:35)
        at org.apache.dolphinscheduler.server.master.engine.WorkflowEventBusFireWorker.doFireSingleEvent(WorkflowEventBusFireWorker.java:144)
        at org.apache.dolphinscheduler.server.master.engine.WorkflowEventBusFireWorker.doFireSingleWorkflowEventBus(WorkflowEventBusFireWorker.java:122)
        ... 8 common frames omitted
```

### What you expected to happen

-

### How to reproduce

set one timeout failure on workflow, and wait for timeout.

### Anything else

_No response_

### Version

dev

### Are you willing to submit PR?

- [x] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

--- a/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/engine/graph/IWorkflowExecutionGraph.java
+++ b/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/engine/graph/IWorkflowExecutionGraph.java
@@ -182,12 +182,12 @@ public interface IWorkflowExecutionGraph {
     /**
      * Whether there exist the TaskExecutionRunnable chain in the graph is finish with paused.
      */
-    boolean isExistPauseTaskExecutionRunnableChain();
+    boolean isExistPausedTaskExecutionRunnableChain();
 
     /**
      * Whether there exist the TaskExecutionRunnable chain in the graph is finish with kill.
      */
-    boolean isExistKillTaskExecutionRunnableChain();
+    boolean isExistKilledTaskExecutionRunnableChain();
 
     /**
      * Check whether the given task is the end of the task chain.

--- a/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/engine/graph/WorkflowExecutionGraph.java
+++ b/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/engine/graph/WorkflowExecutionGraph.java
@@ -204,8 +204,8 @@ public class WorkflowExecutionGraph implements IWorkflowExecutionGraph {
             return false;
         }
         return !isExistFailureTaskExecutionRunnableChain()
-                && !isExistPauseTaskExecutionRunnableChain()
-                && !isExistKillTaskExecutionRunnableChain();
+                && !isExistPausedTaskExecutionRunnableChain()
+                && !isExistKilledTaskExecutionRunnableChain();
     }
 
     @Override
@@ -214,12 +214,12 @@ public class WorkflowExecutionGraph implements IWorkflowExecutionGraph {
     }
 
     @Override
-    public boolean isExistPauseTaskExecutionRunnableChain() {
+    public boolean isExistPausedTaskExecutionRunnableChain() {
         return CollectionUtils.isNotEmpty(pausedTaskChains);
     }
 
     @Override
-    public boolean isExistKillTaskExecutionRunnableChain() {
+    public boolean isExistKilledTaskExecutionRunnableChain() {
         return CollectionUtils.isNotEmpty(killedTaskChains);
     }
 

--- a/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/engine/workflow/statemachine/WorkflowReadyPauseStateAction.java
+++ b/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/engine/workflow/statemachine/WorkflowReadyPauseStateAction.java
@@ -120,13 +120,13 @@ public class WorkflowReadyPauseStateAction extends AbstractWorkflowStateAction {
         }
 
         final WorkflowEventBus workflowEventBus = workflowExecutionRunnable.getWorkflowEventBus();
-        if (workflowExecutionGraph.isExistFailureTaskExecutionRunnableChain()) {
-            workflowEventBus.publish(WorkflowFailedLifecycleEvent.of(workflowExecutionRunnable));
+        if (workflowExecutionGraph.isExistPausedTaskExecutionRunnableChain()) {
+            workflowEventBus.publish(WorkflowPausedLifecycleEvent.of(workflowExecutionRunnable));
             return;
         }
 
-        if (workflowExecutionGraph.isExistPauseTaskExecutionRunnableChain()) {
-            workflowEventBus.publish(WorkflowPausedLifecycleEvent.of(workflowExecutionRunnable));
+        if (workflowExecutionGraph.isExistFailureTaskExecutionRunnableChain()) {
+            workflowEventBus.publish(WorkflowFailedLifecycleEvent.of(workflowExecutionRunnable));
             return;
         }
 

--- a/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/engine/workflow/statemachine/WorkflowReadyStopStateAction.java
+++ b/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/engine/workflow/statemachine/WorkflowReadyStopStateAction.java
@@ -120,7 +120,7 @@ public class WorkflowReadyStopStateAction extends AbstractWorkflowStateAction {
         }
 
         final WorkflowEventBus workflowEventBus = workflowExecutionRunnable.getWorkflowEventBus();
-        if (workflowExecutionGraph.isExistKillTaskExecutionRunnableChain()) {
+        if (workflowExecutionGraph.isExistKilledTaskExecutionRunnableChain()) {
             workflowEventBus.publish(WorkflowStoppedLifecycleEvent.of(workflowExecutionRunnable));
             return;
         }

--- a/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/engine/workflow/statemachine/WorkflowRunningStateAction.java
+++ b/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/engine/workflow/statemachine/WorkflowRunningStateAction.java
@@ -84,7 +84,13 @@ public class WorkflowRunningStateAction extends AbstractWorkflowStateAction {
     public void onStoppedEvent(final IWorkflowExecutionRunnable workflowExecutionRunnable,
                                final WorkflowStoppedLifecycleEvent workflowStoppedEvent) {
         throwExceptionIfStateIsNotMatch(workflowExecutionRunnable);
-        logWarningIfCannotDoAction(workflowExecutionRunnable, workflowStoppedEvent);
+        // [Fix-17354]
+        if (!workflowExecutionRunnable.getWorkflowExecutionGraph().isExistKilledTaskExecutionRunnableChain()) {
+            throw new IllegalStateException(
+                    "The workflow: " + workflowExecutionRunnable.getName()
+                            + " does not exist tasks chain which is killed");
+        }
+        super.workflowFinish(workflowExecutionRunnable, WorkflowExecutionStatus.STOP);
     }
 
     @Override
@@ -142,6 +148,16 @@ public class WorkflowRunningStateAction extends AbstractWorkflowStateAction {
             return;
         }
 
+        // [Fix-17354]
+        // If there exist tasks which has set timeout failed, then will publish a kill event to kill the task.
+        // So there might exist task which is killed, and the workflow instance state is running.
+        // This is a special case, the workflow instance can transform from running to stop state.
+        // Is there better way to handle this case?
+        if (workflowExecutionGraph.isExistKilledTaskExecutionRunnableChain()) {
+            workflowEventBus.publish(WorkflowStoppedLifecycleEvent.of(workflowExecutionRunnable));
+            return;
+        }
+
         if (workflowExecutionGraph.isAllTaskExecutionRunnableChainSuccess()) {
             workflowEventBus.publish(WorkflowSucceedLifecycleEvent.of(workflowExecutionRunnable));
             return;
@@ -149,6 +165,6 @@ public class WorkflowRunningStateAction extends AbstractWorkflowStateAction {
 
         throw new IllegalStateException("The workflow: " + workflowExecutionRunnable.getName() +
                 " state is " + workflowExecutionRunnable.getState()
-                + " can only finish with success/failed but exist task which state is not success and failure");
+                + " can only finish with task success/failed/killed but exist task which state is not success、failure、killed");
     }
 }

--- a/dolphinscheduler-master/src/test/java/org/apache/dolphinscheduler/server/master/integration/WorkflowTestCaseContext.java
+++ b/dolphinscheduler-master/src/test/java/org/apache/dolphinscheduler/server/master/integration/WorkflowTestCaseContext.java
@@ -61,4 +61,14 @@ public class WorkflowTestCaseContext {
         }
         return workflows.get(0);
     }
+
+    public WorkflowDefinition getWorkflow(String name) {
+        if (CollectionUtils.isEmpty(workflows)) {
+            throw new IllegalStateException("workflows is empty");
+        }
+        return workflows.stream()
+                .filter(workflow -> workflow.getName().equals(name))
+                .findFirst()
+                .orElseThrow(() -> new IllegalStateException("Workflow with name " + name + " not found"));
+    }
 }

--- a/dolphinscheduler-master/src/test/java/org/apache/dolphinscheduler/server/master/integration/cases/WorkflowStartTestCase.java
+++ b/dolphinscheduler-master/src/test/java/org/apache/dolphinscheduler/server/master/integration/cases/WorkflowStartTestCase.java
@@ -1004,4 +1004,35 @@ public class WorkflowStartTestCase extends AbstractMasterIntegrationTestCase {
                 });
         masterContainer.assertAllResourceReleased();
     }
+
+    @Test
+    @DisplayName("Test start a workflow which contains a dep task with timeout kill strategy")
+    public void testStartWorkflow_withTimeoutKillTask() {
+        final String yaml = "/it/start/workflow_with_timeout_kill_task.yaml";
+        final WorkflowTestCaseContext context = workflowTestCaseContextFactory.initializeContextFromYaml(yaml);
+        final WorkflowDefinition workflow = context.getWorkflow("workflow_with_timeout_kill_task");
+
+        final WorkflowOperator.WorkflowTriggerDTO workflowTriggerDTO = WorkflowOperator.WorkflowTriggerDTO.builder()
+                .workflowDefinition(workflow)
+                .runWorkflowCommandParam(new RunWorkflowCommandParam())
+                .build();
+        workflowOperator.manualTriggerWorkflow(workflowTriggerDTO);
+
+        await()
+                .atMost(Duration.ofSeconds(90))
+                .untilAsserted(() -> {
+                    Assertions
+                            .assertThat(repository.queryWorkflowInstance(workflow))
+                            .satisfiesExactly(workflowInstance -> assertThat(workflowInstance.getState())
+                                    .isEqualTo(WorkflowExecutionStatus.STOP));
+                    Assertions
+                            .assertThat(repository.queryTaskInstance(workflow))
+                            .hasSize(1)
+                            .satisfiesExactly(taskInstance -> {
+                                assertThat(taskInstance.getName()).isEqualTo("dep_task_with_timeout_killed");
+                                assertThat(taskInstance.getState()).isEqualTo(TaskExecutionStatus.KILL);
+                            });
+                });
+        masterContainer.assertAllResourceReleased();
+    }
 }

--- a/dolphinscheduler-master/src/test/resources/it/start/workflow_with_timeout_kill_task.yaml
+++ b/dolphinscheduler-master/src/test/resources/it/start/workflow_with_timeout_kill_task.yaml
@@ -1,0 +1,95 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+project:
+  name: MasterIntegrationTest
+  code: 1
+  description: This is a fake project
+  userId: 1
+  userName: admin
+  createTime: 2024-08-12 00:00:00
+  updateTime: 2021-08-12 00:00:00
+
+workflows:
+  - name: workflow_with_one_fake_task_success
+    code: 1
+    version: 1
+    projectCode: 1
+    description: This is a fake workflow with single task
+    releaseState: ONLINE
+    createTime: 2024-08-12 00:00:00
+    updateTime: 2021-08-12 00:00:00
+    userId: 1
+    executionType: PARALLEL
+  - name: workflow_with_timeout_kill_task
+    code: 2
+    version: 1
+    projectCode: 1
+    description: This is a fake workflow with single timeout task
+    releaseState: ONLINE
+    createTime: 2024-08-12 00:00:00
+    updateTime: 2021-08-12 00:00:00
+    userId: 1
+    executionType: PARALLEL
+
+tasks:
+  - name: B
+    code: 1
+    version: 1
+    projectCode: 1
+    userId: 1
+    taskType: LogicFakeTask
+    taskParams: '{"localParams":[],"shellScript":"if [ \"${system.project.name}\" = \"MasterIntegrationTest\" ]; then\n  exit 0 \nelse\n  exit 1\nfi","resourceList":[]}'
+    workerGroup: default
+    createTime: 2024-08-12 00:00:00
+    updateTime: 2021-08-12 00:00:00
+    taskExecuteType: BATCH
+  - name: dep_task_with_timeout_killed
+    code: 2
+    version: 1
+    projectCode: 1
+    userId: 1
+    timeoutFlag: 'OPEN'
+    timeoutNotifyStrategy: 'FAILED'
+    timeout: 1
+    taskType: DEPENDENT
+    taskParams: '{"localParams":[],"resourceList":[],"dependence":{"checkInterval":10,"failurePolicy":"DEPENDENT_FAILURE_FAILURE","relation":"AND","dependTaskList":[{"relation":"AND","dependItemList":[{"dependentType":"DEPENDENT_ON_WORKFLOW","projectCode":1,"definitionCode":1,"depTaskCode":0,"cycle":"day","dateValue":"last1Days"}]}]}}'
+    workerGroup: default
+    createTime: 2024-08-12 00:00:00
+    updateTime: 2021-08-12 00:00:00
+    taskExecuteType: BATCH
+
+taskRelations:
+  - projectCode: 1
+    workflowDefinitionCode: 1
+    workflowDefinitionVersion: 1
+    preTaskCode: 0
+    preTaskVersion: 0
+    postTaskCode: 1
+    postTaskVersion: 1
+    createTime: 2024-08-12 00:00:00
+    updateTime: 2024-08-12 00:00:00
+  - projectCode: 1
+    workflowDefinitionCode: 2
+    workflowDefinitionVersion: 1
+    preTaskCode: 0
+    preTaskVersion: 0
+    postTaskCode: 2
+    postTaskVersion: 1
+    createTime: 2024-08-12 00:00:00
+    updateTime: 2024-08-12 00:00:00
+

